### ISSUE TITLE
Open ag-grid directly to a specific page/row that is not page/row 1

### ISSUE THREAD

**Issue by epaitz** [ag-grid/ag-grid#4233]
I am using ag-grid for Angular with a tab control of my own design. The users can open several tables at one time where each table is on its own tab. I am destroying the content of the tab (including the ag-grid) when the tab is not active otherwise the DOM memory grows to large when several tables/tabs are open at the same time. I am using the infinite scrolling model and a server-side data source. 

If the user opens table 1 in tab 1, scrolls down to a row not on the first page, then opens table 2 in tab 2 which means tab 1 for table 1 becomes inactive and table 1 gets destroyed. If the user goes back to tab 1 then table 1 gets recreated and I would like it to open directly to where it was before it was destroyed which would be the row they were at when the table was destroyed. I am specifically describing a scenario where the top row is not on page 1 meaning a second HTTP call had to occur to get the next page of data. 

I am storing the index value for getFirstRenderedRow () from viewportChanged in my app state and setting rowBuffer to 0 so that the index value is visually accurate. This was the suggestion from ag-grid from this GitHub issue. 

Get first row that's actually displayed #1360

When the table gets created, the first or any time, gridReady() fires and I initialize the columnDefs and then call setDatSource(). This will immediately call getRows() on my data source and the params object passed the first time will always be for page 1. I am storing the index value from getFirstRenderedRow() when the table gets destroyed in my app state so my component and my data source know which row index to scroll to when recreating the table. 

I have not been able to find a way to configure ag-grid to initialize directly to a specific row other than row 1 on page 1. The solution I have so far is to wait for the grid to be ready (which means it loads page 1) then from my component call ensureIndexVisible(index, position) using the index number stored in my app state.

The next issue I ran into was how to tell when the grid was ready enough to accept a call to ensureIndexVisible().  The obvious event to use would be firstDataRendered but there appears to be a bug with that even when using server-side row model.

'firstDataRendered' event fires BEFORE data is rendered for grids having rowModelType="serverSide" #2662

After using the setInterval() “hack” described by ag-grid I was able to call ensureIndexVisible() at the correct time so the data grid will scroll to the row the user was on the last time the tab was active. 

While this solution is working it’s not ideal. First, since we are using the server-side row model my client has purchased a license of ag gird and is a paid customer. Since the firstDataRendered event is broken when using the server-side row model that means its broken for all paid customers. Could you please consider fixing this in the next release?  

Second, is there a way to configure ag-grid to open directly to a page/row that is not page/row 1? Having to load page 1 just to get the grid ready enough to then tell it to jump to a specific row (or page) is very inefficient. One idea I had to get around this would be to not only store the last known page in app state but always store page 1 in app state. Then when the table gets recreated (ag-grid will automatically requests page 1) and the page one data will be in memory and fast and there is no need to make a HTTP call then it will go to the last known page/row which will also be in memory.

To sum this up 1.) Consider fixing the firstDataRendered event when using a server-side row model because this is broken for all paid customers. 2.) Provide a way to initialize the data grid to a specific row or page other than row or page 1. If you provide the ability to initialize directly to a specific page/row then I would not have to worry about the firstDataRendered event issue. 

Thanks,
   Eric

**Comment by makinggoodsoftware** [comment]
Hi Eric,

Thanks for the comprehensive summary! Do you have access to our customer support forum? If you don't, you should be able to get it if you contact us at info@ag-grid.com

If you do so, just link to this github, and we can continue the conversation there.

Would that work?

I'll keep an eye here in case you have issues accessing the enterprise forum 
https://ag-grid.zendesk.com/

**Comment by epaitz** [comment]
Are you talking about the zen_desk (or something like that)?

**Comment by makinggoodsoftware** [comment]
Yes, 

As a customer, you should have access to our dedicated customer support forum on https://ag-grid.zendesk.com/

If you are not familiar with this, I would suggest getting in touch with  info@ag-grid.com and just let them know that you need access to the enterprise support channel, you will have to also let them know the license that you are using at the moment in the email

Does this make sense?

**Comment by adonato98** [comment]
Hello @epaitz - I am running into a similar issue on a project I am working on in which I have a use case for initializing a server-side grid to a specific page. Did anything come of this issue in your follow-up with support?

**Comment by epaitz** [comment]
@adonato98  I'm sorry to say we were not able to take this idea any farther than what I've described here. Caching the first page was our solution for not being able to configure ag-grid to load directly to a specific page. This idea became overwhelming complicated because we also support column grouping. If someone grouped a column and that column group had more that one page of records it was difficult to recreate the exact state when the user reactivated the table/tab. 

I wish ad-grid would provide a way to get the exact current state of the table as a JSON string which could be used to recreate or initialize the table to a specific state. The JSON should include the current row/page, selection, sorting, grouping, everything. I'm adding this Incase someone from ag-grid might read this 

Good luck

**Comment by marklagendijk** [comment]
I ran into this issue and solved it by using the [firstDataRendered](https://www.ag-grid.com/javascript-data-grid/grid-events/#reference-gridLifecycle-firstDataRendered) event to set the initial page.

This does mean that you get 2 API requests for any initial page that isn't the first page, but for me this is an acceptable workaround.
